File: app/src/main/java/com/github/teranes10/androidutils/utils/TimerUtil.java
package com.github.teranes10.androidutils.utils;

import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;

import java.util.concurrent.CompletableFuture;

public abstract class TimerUtil {
    private HandlerThread _handlerThread;
    private Handler _handler;
    private Runnable _runnable;

    public abstract String getTag();

    public abstract long getInterval();

    public abstract void execute();

    public void start() {
        if (_handlerThread != null && _handlerThread.isAlive()) {
            return; // Already started
        }

        if (_handlerThread == null) {
            _handlerThread = new HandlerThread(getTag());
            _handlerThread.start();
        }

        if (_handler == null) {
            _handler = new Handler(_handlerThread.getLooper());
        }

        _runnable = () -> CompletableFuture.runAsync(() -> {
            try {
                execute();
            } catch (Exception e) {
                Log.e(getTag(), "execute: ", e);
            }
        }).thenRun(() -> {
            long interval = getInterval();
            if (interval > 0) {
                _handler.postDelayed(_runnable, interval);
            }
        });

        _handler.post(_runnable);
    }

    public void stop() {
        if (_handler != null) {
            _handler.removeCallbacksAndMessages(null);
            _handler = null;
        }
        if (_handlerThread != null) {
            _handlerThread.quit();
            _handlerThread = null;
        }
    }

    public void reset() {
        if (_handler != null && _runnable != null) {
            _handler.removeCallbacksAndMessages(null);
            _handler.postDelayed(_runnable, getInterval());
        }
    }

    public void run() {
        if (_handler != null && _runnable != null) {
            _handler.removeCallbacksAndMessages(null);
            _handler.post(_runnable);
        }
    }
}

File: app/src/main/java/com/github/teranes10/androidutils/utils/TimerUtil.kt
package com.github.teranes10.androidutils.utils

import android.os.SystemClock
import android.util.Log
import com.github.teranes10.androidutils.extensions.CancelableSchedule
import com.github.teranes10.androidutils.extensions.CoroutineScopeExtensions
import com.github.teranes10.androidutils.extensions.CoroutineScopeExtensions.scheduleWithFixedDelay
import com.github.teranes10.androidutils.extensions.ExceptionExtensions.displayMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

abstract class TimerUtil(
    private var interval: Long,
    private val timeUnit: TimeUnit = TimeUnit.MILLISECONDS,
    private val tag: String = "TimerUtil",
    private val executor: ScheduledExecutorService? = null,
    private val scope: CoroutineScope? = null
) {
    private val createdScope = scope ?: CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var cancelableSchedule: CancelableSchedule? = null

    val isRunning: Boolean get() = cancelableSchedule?.isRunning == true

    abstract suspend fun execute()

    fun start(initialDelay: Long = 0) {
        if (interval <= 0 || cancelableSchedule?.isRunning == true) return

        cancelableSchedule = createdScope.scheduleWithFixedDelay(
            delay = interval,
            initialDelay = initialDelay,
            timeUnit = timeUnit,
            executor = executor
        ) {
            execute()
        }
    }

    fun stop() {
        cancelableSchedule?.cancel()
        cancelableSchedule = null
    }

    fun restart(initialDelay: Long = 0) {
        stop()
        start(initialDelay)
    }

    fun updateInterval(newInterval: Long) {
        if (newInterval <= 0 || interval == newInterval) return
        interval = newInterval

        val elapsed = cancelableSchedule?.let { SystemClock.elapsedRealtime() - it.lastTick } ?: 0
        val remaining = (newInterval - elapsed).coerceAtLeast(0)

        restart(remaining)
    }

    fun destroy() {
        stop()

        if (scope == null) {
            createdScope.cancel()
        }
    }
}